test/object-store/c_api/c_api.cpp

```diff
@@ -5800,4 +5801,108 @@ TEST_CASE("app: flx-sync basic tests", "[c_api][flx][sync]") {
+            ws_endpoint.port = endpoint.port;
+            ws_endpoint.path = endpoint.path;
+            for (size_t i = 0; i < endpoint.num_protocols; ++i) {
+                ws_endpoint.protocols.push_back(endpoint.protocols[i]);
+            }
+            ws_endpoint.is_ssl = endpoint.is_ssl;
+
+            m_websocket = socket_provider.connect(realm_websocket_observer->get(), std::move(ws_endpoint));
+        }
+
+        void async_write_binary(util::Span<const char> data, SyncSocketProvider::FunctionHandler&& handler) override
+        {
+            m_websocket->async_write_binary(data, std::move(handler));
+        }
+
+    private:
+        std::unique_ptr<WebSocketInterface> m_websocket;
+    };
+
+    struct TestSyncTimer : realm::c_api::WrapC, SyncSocketProvider::Timer {
+    public:
+        TestSyncTimer(DefaultSocketProvider& socket_provider, std::chrono::milliseconds delay,
+                      SyncSocketProvider::FunctionHandler&& handler)
+        {
+            m_timer = socket_provider.create_timer(delay, std::move(handler));
+        }
+
+        void cancel() override
+        {
+            m_timer->cancel();
+        }
+
+    private:
+        SyncSocketProvider::SyncTimer m_timer;
+    };
+
+    auto logger = std::make_shared<util::StderrLogger>();
+    DefaultSocketProvider default_socket_provider(logger, "SocketProvider");
+
+    auto post_fn = [](realm_userdata_t userdata, realm_sync_socket_callback_t* callback) {
+        auto socket_provider = static_cast<DefaultSocketProvider*>(userdata);
+        auto cb = [callback_copy = callback](Status s) {
+            realm_sync_socket_callback_complete(callback_copy, static_cast<status_error_code_e>(s.code()),
+                                                s.reason().c_str());
+        };
+        socket_provider->post(std::move(cb));
+    };
+    auto create_timer_fn = [](realm_userdata_t userdata, uint64_t delay_ms,
+                              realm_sync_socket_callback_t* callback) -> realm_sync_socket_timer_t {
+        auto socket_provider = static_cast<DefaultSocketProvider*>(userdata);
+        return static_cast<realm_sync_socket_timer_t>(
+            new TestSyncTimer(*socket_provider, std::chrono::milliseconds(delay_ms), std::move(**callback)));
+    };
+    auto cancel_timer_fn = [](realm_userdata_t, realm_sync_socket_timer_t timer_userdata) {
+        auto timer = static_cast<TestSyncTimer*>(timer_userdata);
+        timer->cancel();
+    };
+    auto free_timer_fn = [](realm_userdata_t, realm_sync_socket_timer_t timer_userdata) {
+        realm_release(timer_userdata);
+    };
+    auto websocket_connect_fn =
+        [](realm_userdata_t userdata, realm_websocket_endpoint_t endpoint,
+           realm_websocket_observer_t* realm_websocket_observer) -> realm_sync_socket_websocket_t {
+        auto socket_provider = static_cast<DefaultSocketProvider*>(userdata);
+        return static_cast<realm_sync_socket_websocket_t>(
```

Comment:
For completeness, add a `REQUIRE(socket_provider);` to verify the successful cast.

Reply:
how do you mean "the successful cast" ? how can a static_cast be unsuccessful?

Reply:
What does it do if it's not a pointer to a DefaultSocketProvider? Is this basically just compiler syntax to validate what you're trying to do at compile time and during run time there is really no validation? If it's really a pointer to a Foo object then you're going to be using that as if it's a pointer to a DefaultSocketProvider (with high probability of crashing)?

Reply:
A static_cast doesn't do any checking like that. If you're doing a static_cast to a type that isn't actually the type of the source object, then you're just screwed and your program will crash. I think you're thinking of dynamic_cast which does runtime checking to see if the cast is legal and will return nullptr if it's not. See the first note in https://en.cppreference.com/w/cpp/language/static_cast around downcasting.